core/src/graph/graph.rs

```diff
@@ -0,0 +1,593 @@
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph
+	pub fn import_private<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, keys, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type != self.connection_type {
+			return Err(Error::msg("Incorrect connection type for graph import"))
+		}
+
+		let mut page_map = PageMap::new();
+		for page in pages.iter() {
+			match GraphPage::try_from((page, &keys)) {
+				Err(e) => return Err(e),
+				Ok(p) => {
+					page_map.insert(page.page_id, p);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph, but without decrypting the encrypted portion
+	/// Useful for importing a foreign user's graph for inspection of PRIds
+	pub fn import_opaque<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type.privacy_type() != PrivacyType::Private ||
+			self.connection_type.privacy_type() != PrivacyType::Private
+		{
+			return Err(Error::msg("Invalid privacy type for opaque Private graph import"))
+		}
+
+		let mut page_map = PageMap::new();
+		for page in pages.iter() {
+			page_map.insert(page.page_id, GraphPage::new_opaque(page)?);
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	pub fn calculate_updates<E: EncryptionBehavior>(
+		&self,
+		updates: &Vec<UpdateEvent>,
+		dsnp_user_id: &DsnpUserId,
+		connection_keys: &Vec<DsnpKeys<E>>,
+		encryption_key: (u64, &PublicKey<E>),
+	) -> Result<ExportBundle> {
+		let ids_to_remove: Vec<DsnpUserId> = updates
+			.iter()
+			.filter_map(|event| match event {
+				UpdateEvent::Remove { dsnp_user_id, .. } => Some(*dsnp_user_id),
+				_ => None,
+			})
+			.collect();
+
+		let ids_to_add: Vec<DsnpUserId> = updates
+			.iter()
+			.filter_map(|event| match event {
+				UpdateEvent::Add { dsnp_user_id, .. } => Some(*dsnp_user_id),
+				_ => None,
+			})
+			.collect();
+
+		// First calculate pages that have had connections removed. Later, we will
+		// prefer to use these pages first to add new connections, so as to minimize
+		// the number of pages to update.
+		let pages_with_removals = self.find_connections(&ids_to_remove);
+
+		let mut updated_pages: PageMap = self
+			.pages
+			.iter()
+			.filter_map(|(page_id, page)| {
+				if pages_with_removals.contains(page_id) {
+					let mut updated_page = page.clone();
+					updated_page.remove_connections(&ids_to_remove);
+					return Some((*page_id, updated_page))
+				}
+
+				None
+			})
+			.collect();
+
+		// Now try to add new connections into pages already being updated
+		// Note: these pages have already been cloned, so we don't clone them again
+		let mut add_iter = ids_to_add.iter();
+		while let Some(_) = add_iter.clone().peekable().peek() {
+			if let Some((_page_id, page)) =
+				updated_pages.iter_mut().find(|(_, page)| !page.is_full(false))
+			{
+				let id_to_add = add_iter.next().unwrap();
```

Comment:
is this always safe to do here because of already checked via **`peek`**? a comment might be helpful